nexus/src/db/model.rs

```diff
@@ -352,6 +353,55 @@ pub struct Ipv6Net(pub external::Ipv6Net);
 NewtypeFrom! { () pub struct Ipv6Net(external::Ipv6Net); }
 NewtypeDeref! { () pub struct Ipv6Net(external::Ipv6Net); }
 
+impl Ipv6Net {
+    /// Generate a random subnetwork from this one, of the given prefix length.
+    ///
+    /// `None` is returned if:
+    ///
+    ///  - `prefix` is less than this address's prefix
+    ///  - `prefix` is greater than 128
+    pub fn random_subnet(&self, prefix: u8) -> Option<Self> {
```

Comment:
I think the most important thing is getting the VPC prefix right here with randomness, since that's what'll become important for VPC peering (unless the user specifies something for the VPC at create time) as it'll make it possible to avoid issues there. Within the subnet itself, we should probably optimize for ease for both us and customers. I'm not sure whether that says randomness is good or bad.

Reply:
I think the intent with IPv6 is that they're all just allocating a /64 and fixed size. So this should just be a sequential scan for the first unused integer, probably with a select for update or some other mechanism so it can't change out from under you.

Reply:
I've been thinking more about this. It occurred to me that the CTE is overkill, and that a unique index could be more efficient and easier to maintain. I actually think that's _not_ true, but I wanted to document the logic here. It's useful for me to write these notes out, and to make sure we remember why we're paying the cost of these lines of code.

The point of the query here is to check that the IP address ranges, both v4 and v6, do not conflict, meaning they do not overlap with any other ranges within the VPC. Because all IPv6 subnets are constrained to be `/64`s, they're all the same size, and "overlap" means the same thing as "equal". So we could certainly accomplish the same task with a unique index on `(vpc_id, ipv6_block)`.

For IPv4 ranges, that's not quite true -- two subnets can conflict / overlap, but not be equal: `171.0.0.0/8` and `171.0.0.0/16`, for example. Those overlap, and we want to disallow inserting both, but they're not equal according to CockroachDB. I've gone ahead and tested this in Cockroach:

```
root@127.0.0.1:32221/omicron> CREATE TABLE foo (id INT PRIMARY KEY, net INET);
CREATE TABLE

Time: 4ms total (execution 4ms / network 0ms)

root@127.0.0.1:32221/omicron> CREATE UNIQUE INDEX ON foo (net);
CREATE INDEX

Time: 64ms total (execution 64ms / network 0ms)
```

Then inserting the various records looks like the below. First just insert a record:

```
root@127.0.0.1:32221/omicron> insert into foo VALUES (0, '171.0.0.0/8'::INET);
INSERT 1

Time: 1ms total (execution 1ms / network 0ms)
```

Then insert a record that conflicts on both the `id` and `net` columns. That fails with a `primary` constraint violation:

```
root@127.0.0.1:32221/omicron> insert into foo VALUES (0, '171.0.0.0/8'::INET);
ERROR: duplicate key value violates unique constraint "primary"
SQLSTATE: 23505
DETAIL: Key (id)=(0) already exists.
CONSTRAINT: primary
```

Then inserting something with a new `id`, but the _exact_ same `net` value fails:

```
root@127.0.0.1:32221/omicron> insert into foo VALUES (1, '171.0.0.0/8'::INET);
ERROR: duplicate key value violates unique constraint "foo_net_key"
SQLSTATE: 23505
DETAIL: Key (net)=('171.0.0.0/8') already exists.
CONSTRAINT: foo_net_key
```

Then inserting an overlapping, but not identical subnet works fine. I believe this is the case we want to catch.

```
root@127.0.0.1:32221/omicron> insert into foo VALUES (1, '171.0.0.0/16'::INET);
INSERT 1

Time: 1ms total (execution 1ms / network 0ms)
```

This cannot be caught by a unique index like `(vpc_id, ipv4_block)`, because those networks are not equal. Also note that, [expressions in indexes](https://github.com/cockroachdb/cockroach/issues/9682) recently landed in Cockroach v21.2, but that won't work in our case: the constraint is not a unary function just of the `ipv4_block` column, it's the `inet_contains_or_equals` binary function. So we really do need this query here.

Reply:
Tracking this in #685